orbotservice/src/main/java/org/torproject/android/service/OrbotService.java

```diff
@@ -1487,33 +1338,56 @@ private StringBuffer processSettingsImpl(StringBuffer extraLines) throws IOExcep
             }
         } catch (Exception e) {
             showToolbarNotification(getString(R.string.your_relay_settings_caused_an_exception_), ERROR_NOTIFY_ID, R.drawable.ic_stat_notifyerr);
-
-
             return null;
         }
 
-        ContentResolver mCR = getApplicationContext().getContentResolver();
+        ContentResolver contentResolver = getApplicationContext().getContentResolver();
+        addV3OnionServicesToTorrc(extraLines, contentResolver);
+        addV2HiddenServicesToTorrc(extraLines, contentResolver);
+        addV2ClientCookiesToTorrc(extraLines, contentResolver);
+        return extraLines;
+    }
+
+    private void addV3OnionServicesToTorrc(StringBuffer torrc, ContentResolver contentResolver) {
+        try {
+            Cursor onionServices = contentResolver.query(V3_ONION_SERVICES_CONTENT_URI, V3_ONION_SERVICE_PROJECTION, OnionService.ENABLED + "=1", null, null);
+            if (onionServices != null) {
+                while (onionServices.moveToNext()) {
+                    int localPort = onionServices.getInt(onionServices.getColumnIndex(OnionService.PORT));
+                    int onionPort = onionServices.getInt(onionServices.getColumnIndex(OnionService.ONION_PORT));
+                    String v3DirPath = new File(mV3OnionBasePath.getAbsolutePath(), "v3" + localPort).getCanonicalPath();
+                    torrc.append("HiddenServiceDir ").append(v3DirPath).append("\n");
+                    torrc.append("HiddenServiceVersion 3").append("\n");
+                    torrc.append("HiddenServicePort ").append(onionPort).append(" 127.0.0.1:").append(localPort).append("\n");
+                }
+                onionServices.close();
+            }
+        } catch (Exception e) {
+            Log.e(TAG, e.getLocalizedMessage());
+        }
+    }
```

Comment:
Any chance of doing this configuration via the Control Port?  I suppose that would go against how other things in Orbot have always been handled.  In the native `TorService`, one goal was to make it so that the configuration can be managed in Java then sent to the Control Port as needed.  That means configuration does not require restarting Tor.  If it makes sense, it would be nice to do that now.  But if it means having two parallel modes, somethings written to _torrc_ and some send to the Control Port, it will probably make more sense to put off the Control-Port-only config until `TorService` is integrated.

Reply:
Yes, it would mean that somethings would be sent to the Control Port and others to the initial torrc, but those other things would just be the configuration information for v2 onion services which'll eventually stop being supported anyway. So ya, I think it would make sense to go forward with handling the configuration via the control port

Reply:
Closing this for now ~ regardless of either approach tor needs to be restarted in order to spin up onions